File: lee/src/main/java/org/example/lee/题目/分治/b_排序链表.java
package org.example.lee.题目.分治;

import org.example.lee.model.ListNode;

public class b_排序链表 {

	public ListNode sortList(ListNode head) {
		return tool(head);
	}

	ListNode tool(ListNode node) {
		if (node == null || node.next == null) {
			return node;
		}
		//cut
		ListNode fast = node, slow = node, save;
		while (fast.next != null && fast.next.next != null) {
			slow = slow.next;
			fast = fast.next.next;
		}
		save = slow.next;
		slow.next = null;
		//递归
		ListNode left = tool(node);
		ListNode right = tool(save);
		//merge
		ListNode head = new ListNode(0);
		ListNode save2 = head;
		while (left != null && right != null) {
			if (left.val < right.val) {
				head.next = left;
				left = left.next;
			} else {
				head.next = right;
				right = right.next;
			}
			head = head.next;
		}
		if (left != null) {
			head.next = left;
		} else {
			head.next = right;
		}
		return save2;
	}
}

File: lee/src/main/java/org/example/lee/题目/分治/b_排序链表.go
package 分治

type ListNode struct {
	Val  int
	Next *ListNode
}

// 归并排序
func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	slow, fast, h := head, head, head
	for fast != nil && fast.Next != nil {
		h = slow
		slow = slow.Next
		fast = fast.Next.Next
	}
	//将左右两段分开
	h.Next = nil
	//递归
	right := sortList(slow)
	left := sortList(head)
	//merge
	head = &ListNode{}
	save := head
	for right != nil && left != nil {
		if right.Val > left.Val {
			head.Next = left
			left = left.Next
		} else {
			head.Next = right
			right = right.Next
		}
		head = head.Next
	}
	if right == nil {
		head.Next = left
	} else {
		head.Next = right
	}
	return save.Next
}
